andela/knights-bn-backend
# #170947674 Send notification for the approved request

#### What does this PR do?
- This PR enables a manager to approve the request and then notify the requester via the email and in-app.
#### Description of Task to be completed?
- Create an endpoint for a manager to approve the request
ENDPOINT: ```PATCH http://localhost:4000/api/v1/trips/approve/:requestId```.
- Create an endpoint for a requester to view a specific request.
ENDPOINT: ```GET http://localhost:4000/api/v1/trips/request/:requestId```.
- Send an email notification to the requester about the approved request.
- Notify the requester instantly in the app.
- Test the two endpoints.
- Document the two endpoints.

#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository.
- run `cd knights-bn-backend` to get inside the project.
- run `git fetch && git checkout ft-notification-for-approved-request-170947674` to get inside this branch
- run `npm install` to install all dependencies
- check if you have PostgresQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create2  Postgres databases locally: one for development and another for testing .
- add `BN_EMAIL_NO_REPLY=no-reply@brftnomad.com` in your .env or create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values)

```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'
BN_API_KEY= 'XXXX'
MANAGER_PSWD= 'XXXX'
```
- run `npm run createTables` to create the database tables and seed in primary data.
- run `npm dev-start` to start the server.

1. Log in as a requester to create a new trip
POST : ` http://localhost:4000/api/v1/auth/login`
```
{
    "email": "ishimwewil005@gmail.com",
    "password": "Password@1"
}
```
![image](https://user-images.githubusercontent.com/30593186/76338887-0a2c9600-6302-11ea-8b0d-e5e6179dcccb.png)

2. Copy the token and paste it in the below Codepen code to track in-app notification.
[Codepen link](https://codepen.io/king_Bomaye/pen/VwLbQGX?editors=0011)

Replace the token value with your token in the Codepen.
![image](https://user-images.githubusercontent.com/30593186/76339463-c8501f80-6302-11ea-9d06-1b2ee47d88f8.png)

3. Create a new trip
POST `http://localhost:4000/api/v1/trips/returnTrip`
```
{
	"origin": "Kigali",
	"destination": "Kampala",
	"departureDate": "2020-03-10",
	"returnDate": "2020-03-29",
	"reason": "Having fun",
	"accommodation": "A Hotel"
}
```
![image](https://user-images.githubusercontent.com/30593186/76339808-444a6780-6303-11ea-8d01-e0ab4b2df373.png)

4. log in as an admin
POST : ` http://localhost:4000/api/v1/auth/login`

```
{
    "email": "william.ishimwe@andela.com",
    "password": "Password@1"
}
```
![image](https://user-images.githubusercontent.com/30593186/76339995-8673a900-6303-11ea-836a-3258719737cc.png)

5. View pending requests and pick id of one of the pending request.
GET : ` http://localhost:4000/api/v1/trips/pendingApproval`
![image](https://user-images.githubusercontent.com/30593186/76340103-b15dfd00-6303-11ea-992d-48266ead728f.png)

6.  Approve request.
From the above pending request, let's approve the request with the id of 5. N:B: make sure is the one we pasted his token in the [Codepen link](https://codepen.io/king_Bomaye/pen/VwLbQGX?editors=0011).
![image](https://user-images.githubusercontent.com/30593186/76340387-2f220880-6304-11ea-87c3-76cd1ceb220b.png)

7. Received email:
![image](https://user-images.githubusercontent.com/30593186/76340788-d1da8700-6304-11ea-93d9-22506882ff1d.png)


8. In-app notification:
![image](https://user-images.githubusercontent.com/30593186/76340682-a9eb2380-6304-11ea-99e0-1fd8d9713f5e.png)

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?

[#170947674](https://www.pivotaltracker.com/story/show/170947674)
